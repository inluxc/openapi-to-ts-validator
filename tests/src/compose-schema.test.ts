import path from "node:path";
import fs from "node:fs";
import { generate } from "openapi-typescript-validator";
import Ajv from "ajv";

describe("compose-schema", () => {
  const name = "compose";
  const generatedDir = path.join(__dirname, "../generated", name);
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    if (fs.existsSync(generatedDir))
      fs.rmdirSync(generatedDir, { recursive: true });
    await generate({
      schemaFile: path.join(schemaDir, "compose-schema.js"),
      schemaType: "custom",
      directory: generatedDir,
      decoders: ["BarComponent", "FooComponent", "notfound"],
      standalone: {
        validatorOutput: "module",
      },
    });
  });

  test("schema should match", async () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `schema.json`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
    expect(await new Ajv().validateSchema(JSON.parse(file))).toEqual(true);
  });

  test("files should match", () => {
    const files = fs.readdirSync(generatedDir);
    expect(files).toEqual([
      "decoders",
      "helpers.ts",
      "meta.ts",
      "models.ts",
      "schema.json",
      "validate.ts",
    ]);
  });

  describe("decoders", () => {
    const decodersDir = path.join(generatedDir, `decoders`);

    test("file structure", () => {
      const dir = fs.readdirSync(decodersDir);
      expect(dir).toEqual(["BarComponent", "FooComponent", "index.ts"]);
    });

    test("index.ts", () => {
      const file = fs.readFileSync(path.join(decodersDir, `index.ts`), "utf8");
      expect(file).toMatchSnapshot();
    });
  });

  test("models should match", () => {
    const file = fs.readFileSync(path.join(generatedDir, `models.ts`), "utf8");
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });
});
