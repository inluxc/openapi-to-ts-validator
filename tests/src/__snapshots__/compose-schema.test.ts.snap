// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose-schema decoders index.ts 1`] = `
"/* eslint-disable */

export { BarComponentDecoder } from \"./BarComponent/decoder\";
export { FooComponentDecoder } from \"./FooComponent/decoder\";
"
`;

exports[`compose-schema models should match 1`] = `
"/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export type AnyComponent = BarComponent | FooComponent;

export interface BaseComponent {
  id: string;
}
export interface FooComponent {
  id: string;
  type: \"foo\";
  tag?: number | null;
}
export interface BarComponent {
  id: string;
  type: \"bar\";
  name: string;
}
"
`;

exports[`compose-schema schema should match 1`] = `
"{
  \"type\": \"object\",
  \"title\": \"Schema\",
  \"definitions\": {
    \"BaseComponent\": {
      \"type\": \"object\",
      \"properties\": {
        \"id\": {
          \"type\": \"string\"
        }
      },
      \"required\": [
        \"id\"
      ]
    },
    \"FooComponent\": {
      \"type\": \"object\",
      \"properties\": {
        \"id\": {
          \"type\": \"string\"
        },
        \"type\": {
          \"type\": \"string\",
          \"enum\": [
            \"foo\"
          ]
        },
        \"tag\": {
          \"type\": [
            \"number\",
            \"null\"
          ]
        }
      },
      \"required\": [
        \"id\",
        \"type\"
      ]
    },
    \"BarComponent\": {
      \"type\": \"object\",
      \"properties\": {
        \"id\": {
          \"type\": \"string\"
        },
        \"type\": {
          \"type\": \"string\",
          \"enum\": [
            \"bar\"
          ]
        },
        \"name\": {
          \"type\": \"string\"
        }
      },
      \"required\": [
        \"id\",
        \"type\",
        \"name\"
      ]
    },
    \"AnyComponent\": {
      \"anyOf\": [
        {
          \"$ref\": \"#/definitions/BarComponent\"
        },
        {
          \"$ref\": \"#/definitions/FooComponent\"
        }
      ]
    }
  },
  \"properties\": {
    \"BaseComponent\": {
      \"$ref\": \"#/definitions/BaseComponent\"
    },
    \"FooComponent\": {
      \"$ref\": \"#/definitions/FooComponent\"
    },
    \"BarComponent\": {
      \"$ref\": \"#/definitions/BarComponent\"
    },
    \"AnyComponent\": {
      \"$ref\": \"#/definitions/AnyComponent\"
    }
  }
}"
`;
